snuba/split.py

```diff
@@ -9,86 +9,150 @@
                 else:
-                    overall_result['data'].extend(result['data'])
-
-                if remaining_offset > 0 and len(overall_result['data']) > 0:
-                    to_trim = min(remaining_offset, len(overall_result['data']))
-                    overall_result['data'] = overall_result['data'][to_trim:]
-                    remaining_offset -= to_trim
-
-                total_results = len(overall_result['data'])
-
-                if total_results < limit:
-                    if len(result['data']) == 0:
-                        # If we got nothing from the last query, expand the range by a static factor
-                        split_step = split_step * STEP_GROWTH
-                    else:
-                        # If we got some results but not all of them, estimate how big the time
-                        # range should be for the next query based on how many results we got for
-                        # our last query and its time range, and how many we have left to fetch.
-                        remaining = limit - total_results
-                        split_step = split_step * math.ceil(remaining / float(len(result['data'])))
-
-                    # Set the start and end of the next query based on the new range.
-                    split_end = split_start
-                    try:
-                        split_start = max(split_end - timedelta(seconds=split_step), from_date)
-                    except OverflowError:
-                        split_start = from_date
-
-            return overall_result, status
-        else:
-            return query_func(*args, **kwargs)
+                    # If we got some results but not all of them, estimate how big the time
+                    # range should be for the next query based on how many results we got for
+                    # our last query and its time range, and how many we have left to fetch.
+                    remaining = limit - total_results
+                    split_step = split_step * math.ceil(remaining / float(len(result['data'])))
+
+                # Set the start and end of the next query based on the new range.
+                split_end = split_start
+                try:
+                    split_start = max(split_end - timedelta(seconds=split_step), from_date)
+                except OverflowError:
+                    split_start = from_date
+
+        return overall_result, status
+
+    def col_split(*args, **kwargs):
+        """
+        Split query in 2 steps if a large number of columns is being selected.
+            - First query only selects event_id and project_id.
+            - Second query selects all fields for only those events.
+            - If the orderby term is based on timestamp also update the date range
+        """
+        body = args[0]
+
+        min_cols = ['project_id', 'event_id', 'timestamp']
+
+        minimal_query = {**body, **{'selected_columns': min_cols}}
```

Comment:
The second dict shouldn't be necessary here:

```suggestion
        minimal_query = {**body, 'selected_columns': min_cols}
```